pep-0681.rst

```diff
@@ -0,0 +1,709 @@
+     # non-frozen class, which is an error condition.
+     class Car(Vehicle, frozen=True):
+         wheel_count: int
+
+2. Field ordering and inheritance is assumed to follow the rules
+   specified in PEP 557. This includes the effects of overrides
+   (redefining a field in a child class that has already been defined 
+   in a parent class).
+
+3. PEP 557 indicates that all fields without default values must
+   appear before fields with default values. Although not explicitly
+   stated in PEP 557, this rule is ignored when ``init=False``, and
+   this specification likewise ignores this requirement in that
+   situation. Likewise, there is no need to enforce this ordering when
+   keyword-only parameters are used for ``__init__``, so the rule is
+   not enforced if ``kw_only`` semantics are in effect.
+
+4. As with dataclass, method synthesis is skipped if it would
+   overwrite a method that is explicitly declared within the class.
+   For example, if a class declares an ``__init__`` method explicitly,
+   an ``__init__`` method will not be synthesized for that class.
+
+5. KW_ONLY sentinel values are supported as described in the Python
+   docs [#kw-only-docs]_ and Python issue 43532 [#kw-only-issue]_.
+
+
+Alternate form
+--------------
+
+To avoid delaying adoption of this proposal until after
+``dataclass_transform`` has been added to the ``typing`` module, type
+checkers may support the alternative form ``__dataclass_transform__``.
+This form can be defined locally without any reliance on the
+``typing`` or ``typing_extensions`` modules. It allows immediate
+adoption of this specification by library authors. Type checkers that
+have not yet adopted this specification will retain their current
+behavior.
+
+To use this alternate form, library authors should include the
+following declaration within their type stubs or source files::
+
+  _T = TypeVar("_T")
+  
+  def __dataclass_transform__(
+      *,
+      eq_default: bool = True,
+      order_default: bool = False,
+      kw_only_default: bool = False,
+      field_descriptors: Tuple[Union[type, Callable[..., Any]], ...] = (()),
+  ) -> Callable[[_T], _T]:
+      # If used within a stub file, the following implementation can
+      # be replaced with "...".
+      return lambda a: a
+
+Undefined behavior
+------------------
+
+If multiple ``dataclass_transform`` decorators are found, either on a
+single function/class or within a class hierarchy, the resulting
+behavior is undefined. Library authors should avoid these scenarios.
+
+
+Reference Implementation
+========================
+The pyright [#pyright]_ type checker supports the
+``__dataclass_transform__`` `alternate form`_. Pyright's
+dataClasses.ts source file [#pyright-impl]_ would be a good starting
+point for understanding the implementation.
+
+The attrs [#attrs-usage]_ and pydantic [#pydantic-usage]_ libraries are
+using the ``__dataclass_transform__`` `alternate form`_.
+
+
+Rejected Ideas
```

Comment:
Typically, rejected ideas list at least a brief rationale for _why_ the idea was rejected, as the purpose of this section is to explain to others reading why these ideas were not incorporated into the proposal in its present form. I'd suggest commenting to this effect in the ideas missing it, if possible.